### PR TITLE
Feature/wf 557 c add to cart

### PR DESCRIPTION
**Pull request by geroldboehler**
## Pull request
Adds c-add-to-cart component as well as vuex store. This PR includes 2 Tickets. This branch is based on https://github.com/valantic/vue-template/pull/30 so please review that one first.
 
### Ticket
https://winterhalter-fenner.atlassian.net/browse/WF-557
https://winterhalter-fenner.atlassian.net/browse/WF-558
 
### Browser testing
http://localhost:8080/styleguide/product-detail
 
### Checklist
- [x] Added JSDoc and styleguide demo
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Did run automated tests and linters
- [x] Did assign ticket
- [x] Double checked target branch
 
## Review/Test checklist
- [x] Did run automated tests and linters
- [x] Did review code and documentation
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Re-assign ticket to developer

### CODE REVIEW

**Comment by claudio147** [comment]
- [ ] Check the alignment of the button label-text
bildschirmfoto 2018-05-14 um 10 22 12

- [ ] Add "Stk." to the input field, or is this not part of this ticket?

- [ ] The quantity information "1 Stange = 3m = 0.75kg" is not part of this ticket?

- [ ] The button label should have the color "secondary--1" according to the design, the icon color is correct

File: app/components/c-add-to-cart.vue [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    padding: $spacing--15 $spacing--10 $spacing--30 $spacing--10;
+
+    @include media(xs) {
+      padding: $spacing--0 $spacing--20 $spacing--30 $spacing--20;
+    }
+
+    &__quantity {
+      flex: 0 1 auto;
+      width: 100%;
+      margin-bottom: $spacing--10;
```
Comment (claudio147) [inline app/components/c-add-to-cart.vue]: - [ ] In the design there is a margin-bottom of 15px

File: app/components/c-add-to-cart.vue [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    // render() {},
+  };
+</script>
+
+<style lang="scss">
+  .c-add-to-cart {
+    display: flex;
+    flex-direction: column;
+    padding: $spacing--15 $spacing--10 $spacing--30 $spacing--10;
```
Comment (claudio147) [inline app/components/c-add-to-cart.vue]: - [ ] Don't set outside spacings to the component, set this spacings in the c-product-detail component on a wrapper

File: app/components/c-add-to-cart.vue [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    @include media(xs) {
+      padding: $spacing--0 $spacing--20 $spacing--30 $spacing--20;
```
Comment (claudio147) [inline app/components/c-add-to-cart.vue]: - [ ] same here

File: app/components/c-add-to-cart.vue [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      />
+    </div>
+    <div :class="b('button')">
+      <e-button
+        :progress="progress"
+        primary
+        width="full"
+        @click="onClick"
+      >
+        <e-icon :class="b('icon')" icon="i-cart" inline/><span v-if="hasLabel"> {{ $t('c-add-to-cart.addToCart') }}</span>
```
Comment (claudio147) [inline app/components/c-add-to-cart.vue]: - [ ] Don't add bem class to a included component, use  a wrapper instead

File: app/components/c-product-detail.vue [hunk 5]
```diff
@@ -22,7 +22,8 @@
           <div :class="b('prices')">
             <c-prices :price-gross="erp.priceGross" :price="erp.price"/>
           </div>
-          / qty / add to cart
+          / qty /
```
Comment (claudio147) [inline app/components/c-product-detail.vue]: - [ ] Can also get removed when c-add-to-cart component includes the quantity

File: app/components/c-add-to-cart.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+Renders add to cart component.
```
Comment (patric-eberle) [inline app/components/c-add-to-cart.md]: - [ ] Please move description to top of `export default {}` of component itself.

File: app/components/c-add-to-cart.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+#### default
+```vue
+<c-add-to-cart sku="12345"/>
+```
+
+#### default with label
+```vue
+<c-add-to-cart sku="12345" label/>
+```
```
Comment (patric-eberle) [inline app/components/c-add-to-cart.md]: - [ ] Please add `step` example

File: app/components/c-add-to-cart.vue [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    // mounted() {},
+    // beforeUpdate() {},
+    // updated() {},
+    // activated() {},
+    // deactivated() {},
+    // beforeDestroy() {},
+    // destroyed() {},
+
+    methods: {
+      ...mapActions({
```
Comment (patric-eberle) [inline app/components/c-add-to-cart.vue]: - [ ] I would recommend to use "named mappers" as it removes code duplication and gives a better overview of mapped stuff:

```
methods: {
  ...mapActions('cart', [
    'addToCart'
  ])
}
```

See also https://vuex.vuejs.org/en/modules.html#binding-helpers-with-namespace

File: app/components/c-add-to-cart.vue [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        addToCart: 'cart/addToCart',
+      }),
+      onClick() {
+        this.progress = true;
+
+        this.addToCart(this.sku, this.quantity)
+          .then((response) => {
+            this.progress = false;
```
Comment (patric-eberle) [inline app/components/c-add-to-cart.vue]: - [ ] Please use `.finally()` for data resets after a promise.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

File: app/store/modules/cart/index.js [hunk 10]
```diff
@@ -1,10 +1,55 @@
...
+     * @param {string} quantity   Amount of product units to be added
+     * @param {string} sku   Sku of the product to be added
+     *
+     * @returns {Promise} Promise object
+     */
+    addToCart({ commit }, sku, quantity) {
+      return api.post('/cart/1', { sku, quantity }) // TODO - replace id of cart
+        .then((response) => {
+          if (response && response.data && response.data.totals && Object.keys(response.data.totals).length) {
+            commit('updateTotals', response.data.totals);
```
Comment (patric-eberle) [inline app/store/modules/cart/index.js]: - [ ] Shouldn't this replace the whole cart data?

File: app/store/utils/api.js [hunk 11]
```diff
@@ -30,4 +30,17 @@ export default {
       .then(response => Promise.resolve(response))
       .catch(error => Promise.reject(error));
   },
+
+  /**
+   * Runs a patch request with a given url and payload.
+   * @param {*} url url to patch to
```
Comment (patric-eberle) [inline app/store/utils/api.js]: - [ ] Please use blank lines when starting a new documentation section (`@param`, `@returns`)

Comment (patric-eberle) [inline app/store/utils/api.js]: - [ ] Please define type for url (guess its `string`)

File: app/store/utils/api.js [hunk 12]
```diff
@@ -30,4 +30,17 @@ export default {
...
+   * @param {*} payload patch payload which will be attached to the request
```
Comment (patric-eberle) [inline app/store/utils/api.js]: - [ ] Please define the type for `payload` (guess its `object`).

File: app/components/c-add-to-cart.vue (hunk 9 above)
Comment (geroldboehler) [reply app/components/c-add-to-cart.vue]: Finally doesn't receive a parameter which i need for debug output. I have to refactor this anyway in WF-600-mini-cart

File: app/store/modules/cart/index.js (hunk 10 above)
Comment (geroldboehler) [reply app/store/modules/cart/index.js]: The POST request only returns the totals, not the items

**Comment by geroldboehler** [comment]
I was not able to prevent character input on ie 11 / safari. However, i will strip the chars before submitting them to the api.

File: app/components/c-add-to-cart.vue (hunk 9 above)
Comment (patric-eberle) [reply app/components/c-add-to-cart.vue]: @geroldboehler Handling the response in `then()` is ok. But `this.progress = false` should be handled inside `finally()` to make sure it's always executed.

**Pull request merged**